### PR TITLE
Core/Creature: Implement vendor item count and restock timer storage

### PR DESCRIPTION
**Pull request by robinsch**
**Changes proposed:**
-  Load and save current vendor item count and restock timer
  (In case of a server crash all limited items at vendor refilled.)

**Target branch(es):** 3.3.5/6.x

**Tests performed:** (Does it build, tested in-game, etc)
First 3 commits were tested in-game. 4th is a WIP.

**Known issues and TODO list:**
- [ ] Renaming?
- [ ] Feedback about (bad) code?

SQL:

``` sql
CREATE TABLE `creature_vendor` (
  `spawnId` int(10) unsigned NOT NULL COMMENT 'Global Unique Identifier',
  `instanceId` int(10) unsigned NOT NULL DEFAULT '0',
  `item` mediumint(8) unsigned NOT NULL DEFAULT '0',
  `count` tinyint(3) unsigned NOT NULL DEFAULT '0',
  `lastIncrementTime` int(10) unsigned NOT NULL DEFAULT '0',
  PRIMARY KEY (`spawnId`,`instanceId`)
);
```

### CODE REVIEW

**Comment by Shauren** [comment]
World database is read-only.

**Comment by Aokromes** [comment]
If it's temp data it must be on characters database.

**Comment by Aokromes** [comment]
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Globals/ObjectMgr.cpp:9316:48: fatal error: 

      no matching constructor for initialization of 'VendorItemCount'

  ...VendorItemCount(item, count, lastIncrementTime));

     ^               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/home/travis/build/TrinityCore/TrinityCore/src/server/game/Entities/Creature/Creature.h:375:14: note: 

      candidate constructor not viable: requires 2 arguments, but 3 were

      provided

    explicit VendorItemCount(uint32 _item, uint32 _count)

             ^

/home/travis/build/TrinityCore/TrinityCore/src/server/game/Entities/Creature/Creature.h:373:8: note: 

      candidate constructor (the implicit copy constructor) not viable: requires

      1 argument, but 3 were provided

struct VendorItemCount

       ^

/home/travis/build/TrinityCore/TrinityCore/src/server/game/Entities/Creature/Creature.h:373:8: note: 

      candidate constructor (the implicit move constructor) not viable: requires

      1 argument, but 3 were provided

/home/travis/build/TrinityCore/TrinityCore/src/server/game/Entities/Creature/Creature.cpp:233:13: fatal error: 

      use of undeclared identifier 'LoadVendorItemCount'

            LoadVendorItemCount();

            ^

File: src/server/game/Globals/ObjectMgr.cpp [hunk 1]
```diff
@@ -9285,3 +9285,37 @@ void ObjectMgr::LoadCreatureQuestItems()
...
+void ObjectMgr::LoadVendorItemCounts()
+{
+    uint32 oldMSTime = getMSTime();
+
+    // For reload case
+    for (VendorItemCountsContainer::iterator itr = _vendorItemCountsStore.begin(); itr != _vendorItemCountsStore.end(); ++itr)
+        itr->second.clear();
+    _vendorItemCountsStore.clear();
+
+    QueryResult result = WorldDatabase.Query("SELECT guid, item, count, lastIncrementTime FROM creature_vendor");
```
Comment (Shauren) [inline src/server/game/Globals/ObjectMgr.cpp]: Wrong database

File: src/server/game/Globals/ObjectMgr.cpp [hunk 2]
```diff
@@ -9285,3 +9285,37 @@ void ObjectMgr::LoadCreatureQuestItems()
...
+    if (!result)
+    {
+        TC_LOG_ERROR("sql.sql", ">>  Loaded 0 Vendor Item Counts. DB table `creature_vendor` is empty!");
```
Comment (Shauren) [inline src/server/game/Globals/ObjectMgr.cpp]: Not having any saved data is not an error, do not log it with TC_LOG_ERROR

File: src/server/game/Globals/ObjectMgr.h [hunk 3]
```diff
@@ -671,6 +671,8 @@ typedef std::unordered_map<uint32, DungeonEncounterList> DungeonEncounterContain
 
 class PlayerDumpReader;
 
+typedef std::unordered_map<uint32, VendorItemCounts> VendorItemCountsContainer;
```
Comment (Shauren) [inline src/server/game/Globals/ObjectMgr.h]: Map key is the spawn id, use proper type here ObjectGuid::LowType, not uint32

File: src/server/game/Globals/ObjectMgr.cpp [hunk 4]
```diff
@@ -9285,3 +9285,37 @@ void ObjectMgr::LoadCreatureQuestItems()
...
+        return;
+    }
+
+    uint32 count = 0;
+
+    do
+    {
+        Field* fields = result->Fetch();
+
+        uint64 guid = fields[0].GetUInt32();
```
Comment (Shauren) [inline src/server/game/Globals/ObjectMgr.cpp]: Map key is the spawn id, use proper type here ObjectGuid::LowType, not uint32

**Comment by Shauren** [comment]
Will fail horribly when vendors with limited stock are spawned inside instances, spawn id will not be unique

**Comment by robinsch** [comment]
@Shauren: About the second part - I don't really understand that. The state will be updated at GetVendorItemCurrentCount and UpdateVendorItemCurrentCount (whenever somebody talkes to a vendor or buys something from him)

**Comment by Shauren** [comment]
Indeed you did not understand.
What will happen is
* Startup, ObjectMgr::LoadVendorItemCounts() (STEP 1)
* Loading grid, Creature::LoadVendorItemCount(), pulls data loaded during STEP 1
* Player buys something from the vendor, calling Creature::SaveVendorItemCount() which updates it in database
* Unloading grid because of no players in it
* Loading grid, new creature is spawned, calling Creature::LoadVendorItemCount(), pulls data loaded during STEP 1 (the initial startup count!)

**Comment by robinsch** [comment]
I see, thanks.

**Comment by Shauren** [comment]
Not threadsafe at all. Store this in Map, not in ObjectMgr
DB key should be (instanceid, spawnid) pair

**Comment by robinsch** [comment]
Okay, will update this tomorrow. Thanks!

File: src/server/game/Entities/Creature/Creature.cpp [hunk 5]
```diff
@@ -2926,3 +2930,28 @@ void Creature::ClearTextRepeatGroup(uint8 textGroup)
...
+{
+    sObjectMgr->SaveVendorItemCounts(GetSpawnId(), vendorItemCount);
+        
+    SQLTransaction trans = CharacterDatabase.BeginTransaction();
+
+    PreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_REP_CREATURE_VENDOR);
+    uint8 index = 0;
+
+    stmt->setUInt32(index++, m_spawnId);
+    stmt->setUInt32(index++, itr->itemId);
```
Comment (tkrokli) [inline src/server/game/Entities/Creature/Creature.cpp]:
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Entities/Creature/Creature.cpp:2951:30: fatal error: 
      use of undeclared identifier 'itr'
    stmt->setUInt32(index++, itr->itemId);
                             ^
```

**Comment by robinsch** [comment]
@Shauren Moved to map, where should I call load? Map::AddToMap(T* obj)

**Comment by tkrokli** [comment]
Isn't it strange how much the AppVeyor and Travis CI builds fail because of Unit.h not defining WorldPacket and other definitions being wrong or missing? I can't see why the changes in this PR should trigger those error messages, but I have probably overlooked something. Any ideas or thoughts?

File: src/server/database/Database/Implementation/CharacterDatabase.cpp [hunk 6]
```diff
@@ -326,6 +326,10 @@ void CharacterDatabaseConnection::DoPrepareStatements()
     PrepareStatement(CHAR_DEL_CREATURE_RESPAWN_BY_INSTANCE, "DELETE FROM creature_respawn WHERE mapId = ? AND instanceId = ?", CONNECTION_ASYNC);
     PrepareStatement(CHAR_SEL_MAX_CREATURE_RESPAWNS, "SELECT MAX(respawnTime), instanceId FROM creature_respawn WHERE instanceId > 0 GROUP BY instanceId", CONNECTION_SYNCH);
 
+    // Creature
+    PrepareStatement(CHAR_SEL_CREATURE_VENDOR, "SELECT spawnId, item, count, lastIncrementTime FROM corpse WHERE instanceId = ?", CONNECTION_ASYNC);
```
Comment (robinsch) [inline src/server/database/Database/Implementation/CharacterDatabase.cpp]: typo

**Comment by robinsch** [comment]
@tkrokli It's because of #include "Creature.h". Fixed with next update.

File: src/server/game/Maps/Map.h [hunk 7]
```diff
@@ -31,6 +31,7 @@
 #include "DynamicTree.h"
 #include "GameObjectModel.h"
 #include "ObjectGuid.h"
+#include "Creature.h"
```
Comment (Shauren) [inline src/server/game/Maps/Map.h]: you dont even need this here - for the future please carefully think about adding includes inside other headers

Comment (Aokromes) [inline src/server/game/Maps/Map.h]: @robinsch

**Comment by Aokromes** [comment]
continuous-integration/appveyor/pr — AppVeyor build failed
Details continuous-integration/travis-ci/pr — The Travis CI build failed

@robinsch look at 

https://github.com/TrinityCore/TrinityCore/commit/e2c01899272d1e6c6bbf990676b818cceaf18c9c

to learn how to handle updates on auth/characters databases :)

**Comment by robinsch** [comment]
I will update this when I find some time.

**Pull request closed without merging**